lib/widgets.js

```diff
@@ -43,6 +44,82 @@ var input = function (type) {
     };
 };
 
+var choiceValueEquals = function (value1, value2) {
+    return !is.array(value1) && !is.array(value2) && String(value1) === String(value2);
+};
+
+var isSelected = function (value, choice) {
+    return value && (is.array(value) ? value.some(choiceValueEquals.bind(null, choice)) : choiceValueEquals(value, choice));
+};
+
+var renderChoices = function (choices, renderer) {
+    return choices.reduce(function (partialRendered, choice) {
+        var isNested = is.array(choice[1]);
+        var renderData = isNested ?
+            { isNested: true, label: choice[0], choices: choice[1] } :
+            { isNested: false, value: choice[0], label: choice[1] };
+        return partialRendered + renderer(renderData);
+    }, '');
+};
+
+var isScalar = function (value) {
+    return is.string(value) || is.number(value) || is.bool(value);
+};
+
+var validateChoices = function (choices) {
+    choices.forEach(function (choice) {
+        if (choice.length !== 2) {
+            throw new Error('choice array must have two elements');
+        }
+
+        if (isScalar(choice[0]) && isScalar(choice[0])) {
+            return;
+        } else if (isScalar(choice[0]) && is.array(choice[1])) {
+            validateChoices(choices[1]);
```

Comment:
in and of itself yes. validateChoices has no safeguards against infinite recursion. but it’s only called once choices have been converted to the internal array-based format and the maximum depth for any widget is 1. as the conversion throws an exception once the maximum depth is exceeded i can’t see how to provoke an infinite recursion. but i can add a heads up as documentation so no one else uses this function without taking care of recursion before.